### PR TITLE
batch raycaster-intersection events, reuse event detail

### PR DESCRIPTION
**Pull request by ngokevin**
**Description:**

**Changes proposed:**
- Cache event detail object.
- Emit one raycaster-intersection event per tick for all intersected entities.

### CODE REVIEW

**Comment by machenmusik** [comment]
(This may break existing code, right?)

**Comment by vincentfretin** [comment]
Yes definitely, this will break super-hands for GearVR. super-hands is already broken with the merge of #3126

**Comment by dmarcos** [comment]
@vincentfretin Is it too hard for `super-hands` to accommodate to the new API?

**Comment by ngokevin** [comment]
Yeah, didn't see super-hands coming. I can try making a small PR.

**Comment by machenmusik** [comment]
This is one of those features where it is rarely used, but for any that do, the event detail is always used at the moment. Would it be feasible to determine if there was anything listening and if not just don’t emit at all?

**Comment by ngokevin** [comment]
Updated to use @dmarcos' cached event pattern.

**Comment by machenmusik** [comment]
Nice!

File: tests/components/raycaster.test.js [hunk 1]
```diff
@@ -207,16 +207,6 @@ suite('raycaster', function () {
...
-      var raycasterEl = el;
-      raycasterEl.addEventListener('raycaster-intersection', function (evt) {
-        assert.equal(evt.detail.els[0], targetEl);
-        assert.equal(evt.detail.intersections[0].object.el, targetEl);
-        done();
-      });
-      component.tick();
-    });
-
     test('emits event on intersected entity with details', function (done) {
```
Comment (vincentfretin) [inline tests/components/raycaster.test.js]: You should put back this test, no?

**Comment by vincentfretin** [comment]
@dmarcos @ngokevin I made a small PR to super-hands https://github.com/wmurphyrd/aframe-super-hands-component/pull/80

File: src/components/raycaster.js [hunk 2]
```diff
@@ -148,98 +151,84 @@ module.exports.Component = registerComponent('raycaster', {
...
-      // Emit intersected on intersected entity per intersected entity.
-      for (i = 0; i < intersections.length; i++) {
-        intersections[i].object.el.emit('raycaster-intersected', {
-          el: el,
-          intersection: intersections[i]
-        });
-      }
+    // Emit intersected on intersected entity per intersected entity.
+    for (i = 0; i < intersections.length; i++) {
+      intersections[i].object.el.emit('raycaster-intersected', {
```
Comment (dmarcos) [inline src/components/raycaster.js]: we can use `intersectionDetail` for this one too

File: src/components/cursor.js [hunk 3]
```diff
@@ -233,9 +233,9 @@ module.exports.Component = registerComponent('cursor', {
     var intersection;
 
     // Select closest object, excluding the cursor.
-    index = evt.detail.els[0] === cursorEl ? 1 : 0;
-    intersection = evt.detail.intersections[index];
-    intersectedEl = evt.detail.els[index];
+    index = cursorEl.components.raycaster.intersectedEls[0] === cursorEl ? 1 : 0;
```
Comment (dmarcos) [inline src/components/cursor.js]: temp variable `var raycasterComponent = cursorEl.components.raycaster`?

File: src/components/raycaster.js (hunk 2 above)
Comment (dmarcos) [reply src/components/raycaster.js]: We probably need an `intersectedDetail` object

Comment (ngokevin) [reply src/components/raycaster.js]: This one is different because it's different data per intersected entity.

**Comment by ngokevin** [comment]
Updated.

**Comment by ngokevin** [comment]
This has been done as part of another one of my PRs https://github.com/aframevr/aframe/pull/3250

**Pull request closed without merging**